pkg/api/svc.go

```diff
@@ -98,6 +129,88 @@ func (s *Service) Run(ctx context.Context) error {
 	return nil
 }
 
+// getRequestLimits is a helper function transform request limits mentioned in server configuration
+// into wasabi request limits, so that we can use the RateLimiter middleware.
+func getRequestLimits(rateLimitCfg RateLimits) (func(wasabi.Request) (string, time.Duration, uint64), error) {
+	if len(rateLimitCfg.Groups) == 0 {
+		return getDefaultRequestLimits(rateLimitCfg.General)
+	}
+
+	return getRateLimitForMethods(rateLimitCfg)
+}
+
+func getRateLimitForMethods(rateLimitCfg RateLimits) (func(wasabi.Request) (string, time.Duration, uint64), error) {
+	groupRatesMap, err := buildGroupRateMap(rateLimitCfg.Groups)
+
+	if err != nil {
+		return nil, err
+	}
+
+	generalRateLimitFunc, err := getDefaultRequestLimits(rateLimitCfg.General)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return func(r wasabi.Request) (string, time.Duration, uint64) {
+		group, exists := groupRatesMap[r.RoutingKey()]
+
+		if !exists {
+			return generalRateLimitFunc(r)
+		}
+
+		ip := getIPFromRequest(r)
+
+		if duration, err := time.ParseDuration(group.Limits.Interval); err == nil {
+			return ip, duration, uint64(group.Limits.Limit)
+		}
```

Comment:
first return value should be `group name` + `ip` to rate limit correctly for the group.
also duration parsing should be done during initialization, not at run time

Reply:
duration parsing cannot be done prior to request as we need to lookup the routing key from request to find the correct group. Maybe I can refactor this by keeping the groups mapped to duration already